src/D2L.CodeStyle.Analyzers/Async/BlockingAnalyzer.cs

```diff
@@ -0,0 +1,359 @@
+
+				ctx.ReportDiagnostic(
+					Diagnostic.Create(
+						Diagnostics.AsyncMethodCannotBeBlocking,
+						attrData.ApplicationSyntaxReference.GetSyntax().GetLocation(),
+						methodSymbol.Name
+					)
+				);
+				return;
+			}
+
+			// Don't let us use [Blocking] if we're implementing something non-blocking,
+			// callers need to know that they may block.
+			if( implementedNonBlockingThings.Count > 0 ) {
+				// TODO: code fix to remove [Blocking]
+
+				ctx.ReportDiagnostic(
+					Diagnostic.Create(
+						Diagnostics.DontIntroduceBlockingInImplementation,
+						attrData.ApplicationSyntaxReference.GetSyntax().GetLocation(),
+						$"{methodSymbol.ContainingType.Name}.{methodSymbol.Name}",
+						// Just use the first one as an example:
+						$"{implementedNonBlockingThings[0].ContainingType.Name}.{implementedNonBlockingThings[0].Name}"
+					)
+				);
+				return;
+			}
+
+			Debug.Assert(
+				// We are [Blocking]
+				attrData != null
+				// everything we implemement is too
+				&& implementedNonBlockingThings.Count == 0
+			);
+		}
+
+		private static void InspectMethodCall(
+			SyntaxNodeAnalysisContext ctx,
+			INamedTypeSymbol blockingAttr,
+			Action<IMethodSymbol> noteMethodHasBlockingCall
+		) {
+			// Goal: find calls to blocking methods inside non-blocking methods, or in properties.
+
+			var invocation = (InvocationExpressionSyntax)ctx.Node;
+
+			var invokedSymbolInfo = ctx.SemanticModel.GetSymbolInfo( invocation.Expression );
+
+			// Unfortunately there are ways to use this to dodge analysis, but
+			// there is only so much we can do.
+			if( invokedSymbolInfo.Symbol?.Kind != SymbolKind.Method ) {
+				return;
+			}
+
+			var invokedMethodSymbol = (IMethodSymbol)invokedSymbolInfo.Symbol;
+
+			// async methods are never [Blocking] (see other parts of the analyzer)
+			if( invokedMethodSymbol.IsAsync ) {
+				return;
+			}
+
+			// If the thing we're calling isn't [Blocking] then OK.
+			// TODO: support other means of inferring blockingness for 3rd party libraries.
+			if( !HasAttribute( invokedMethodSymbol, blockingAttr ) ) {
+				return;
+			}
+
+			if( !TryGetContainingMethod( ctx, invocation, invokedMethodSymbol, out var myMethodDeclaration ) ) {
+				return;
+			}
+
+			var myMethodSymbol = (IMethodSymbol)ctx.SemanticModel.GetDeclaredSymbol( myMethodDeclaration );
+
+			if( myMethodSymbol == null ) {
+				return;
+			}
+
+			if( HasAttribute( myMethodSymbol, blockingAttr ) ) {
+				// We have [Blocking] so we're allowed to call other [Blocking] things,
+				// but make note that we definitely require blocking.
+				noteMethodHasBlockingCall( myMethodSymbol );
+				return;
+			}
+
+			if( myMethodSymbol.IsAsync ) {
+				ctx.ReportDiagnostic(
+					Diagnostic.Create(
+						Diagnostics.AsyncMethodCannotCallBlockingMethod,
+						invocation.GetLocation(),
+						invokedMethodSymbol.Name
+					)
+				);
+
+				// TODO: code fix that looks for an async equivalent and suggests calling it instead.
+				return;
+			}

```

Comment:
Same question as above, re: returning awaitables/tasklikes

```csharp
public Task FooAsync() {
  BlockingCall();
  BlockingCall2();

  return BarAsync();
}
```

Reply:
ditto -- I intentionally allowed blocking in here because it isn't itself trouble, but it is debatable.